[FLINK-16245] Decoupling user classloader from context classloader.

AHeise opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

This is a follow-up PR of https://github.com/apache/flink/pull/11303 after https://github.com/apache/flink/pull/11611 disabled it because of performance regression in blink e2e test.

## What is the purpose of the change

Decoupling user class loader from context classloader.

Thus, user classloader can be unloaded even though a reference on the context classloader outlives the user code.

## Brief change log

- Adding SafetyNetWrapperClassLoader
- Use it for FlinkUserCodeClassLoaders
- Make sure LocalExecutor closes user classloader

Compared to the original PR, it includes the following changes
- Use the parent classloader of the FlinkUserCodeClassLoader in SafetyNetWrapperClassLoader to address the performance concerns.
- Further, improve performance by enabling parallel class loading in all involved classloaders.

## Verifying this change

Added unit test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zentol commented:
Could you summarize what the problem was that caused the performance regression?

AHeise commented:
> Could you summarize what the problem was that caused the performance regression?

For some reason, not setting a parent classloader in the wrapper is much slower. I cannot see an obvious reason and added a couple of measurements.

First, I wanted to know if the performance difference was caused by slower class loading. So I measured the total time spent in `ClassLoader#loadClass` for the previous version and this version. I found no difference. There is also no difference by the amount of classes loaded.

Second, a profiler revealed that much time in the slow version was spent in `Class#newInstance`. I added the time measurements to `org.apache.flink.table.runtime.generated.GeneratedClass`. Interestingly, there is again no difference in compiling the class. There is a huge difference in calling `newInstance` on the generated class though that can easily account for the time difference. 

Without the parent, creating an instance of the 4 specific blink operators `BatchNestedLoopJoin`, `LongHashJoinOperator`, `LocalHashAggregateWithKeys`, and `HashAggregateWithKeys` slows down over time taking up over 10s to finish at the end of the. 

Interestingly, they use only the instantiation method with explicit arguments `GeneratedClass#newInstance(ClassLoader classLoader, Object... args)`. However, there are also operators that use that method that are "well-behaved".

I have also forced class resolution through reflection on the generated classes, but resolution is fast in all cases. It's really just about creating the instances.

AHeise commented:
There was also a [classloader leak in ORC](https://issues.apache.org/jira/browse/ORC-653), for which I added a workaround.

zentol commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSink.java:
```diff
@@ -241,12 +242,14 @@ public final DataStreamSink consumeDataStream(DataStream dataStream) {
 			return Optional.of(ParquetRowDataBuilder.createWriterFactory(
 					formatType, formatConf, hiveVersion.startsWith("3.")));
 		} else if (serLib.contains("orc")) {
+			Configuration formatConf = new ThreadLocalClassLoaderConfiguration(jobConf);
```
If we were to use a normal configuration, what would happen? Subsequent jobs would not be able to run?

AHeise replied:
Yes. That makes me believe that we should probably make the safety net configurable (on by default). Or would you think we can catch up and fix leaks much quicker?

zentol replied:
An opt-out switch is a good idea; also for cases where things don't behave as expected close to the next release, and we don't want to revert commits at that point.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java:
```diff
@@ -124,6 +124,15 @@
 
+	@Documentation.Section(Documentation.Sections.EXPERT_CLASS_LOADING)
+	public static final ConfigOption<Boolean> CHECK_LEAKED_CLASSLOADER = ConfigOptions
+		.key("classloader.check_leaked_classloader")
```
we don't use underscores.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java:
```diff
@@ -136,7 +150,10 @@ public void close() {
+				throw new IllegalStateException("Trying to access closed classloader. Please check if you store " +
+					"classloaders directly or indirectly in static fields. If the stacktrace suggests that the leak " +
+					"occurs in a third party library and cannot be fixed immediately, you can disable this check " +
+					"with the configuration 'classloader.check_leaked_classloader'.");
```
Retrieve the key from the actual configuration object instead.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java:
```diff
@@ -124,6 +124,15 @@
+		.key("classloader.check_leaked_classloader")
+		.booleanType()
+		.defaultValue(true)
+		.withDescription("Fails if a user classloader is used in another job after a job has been finished. This " +
+			"check should only be disabled if the class leak occurs in an external library cannot be fixed " +
+			"immediately.");
```
```
Fails attempts at loading classes if the user classloader of a job is used after it has terminated.
This is usually caused by the classloader being leaked by lingering threads or misbehaving libraries, which may also result in the classloader being used by other jobs.
This check should only be disabled if such a leak prevents further jobs from running.
```
